pkg/suggestion/nasrl_service.py

```diff
@@ -1,22 +1,227 @@
-import random
-import string
-
+from pkg.suggestion.NAS_Reinforcement_Learning.Controller import Controller
+from pkg.suggestion.NAS_Reinforcement_Learning.Operation import SearchSpace
+from pkg.suggestion.NAS_Reinforcement_Learning.SuggestionParam import parseSuggestionParam
+import tensorflow as tf
 import grpc
-import numpy as np
-
 from pkg.api.python import api_pb2
 from pkg.api.python import api_pb2_grpc
 import logging
 from logging import getLogger, StreamHandler, INFO, DEBUG
+import json
+import os
 
 
 class NasrlService(api_pb2_grpc.SuggestionServicer):
-    def __init__(self, logger=None):
+    def __init__(self):
         self.manager_addr = "vizier-core"
         self.manager_port = 6789
+        self.current_study_id = ""
         self.current_trial_id = ""
+        self.ctrl_cache_file = ""
+        self.ctrl_step = 0
+        self.tf_graph = tf.get_default_graph()
+        self.is_first_run = True
+
+    def reset_controller(self, request):
+        print("-" * 80 + "\nResetting Suggestion for StudyJob {}\n".format(request.study_id) + "-" * 80)
+        self.ctrl_step = 0
+        self.current_study_id = request.study_id
+        self.ctrl_cache_file = "ctrl_cache/{}.ckpt".format(self.current_study_id)
+        if not os.path.exists("ctrl_cache/"):
+            os.makedirs("ctrl_cache/")
+        self.current_trial_id = ""
+        self._get_suggestion_param(request.param_id)
+        self._get_search_space(request.study_id)
+
+        with self.tf_graph.as_default():
+            self.controller = Controller(
+                num_layers=self.num_layers,
+                num_operations=self.num_operations,
+                lstm_size=self.suggestion_config['lstm_num_cells'],
+                lstm_num_layers=self.suggestion_config['lstm_num_layers'],
+                lstm_keep_prob=self.suggestion_config['lstm_keep_prob'],
+                lr_init=self.suggestion_config['init_learning_rate'],
+                lr_dec_start=self.suggestion_config['lr_decay_start'],
+                lr_dec_every=self.suggestion_config['lr_decay_every'],
+                lr_dec_rate=self.suggestion_config['lr_decay_rate'],
+                l2_reg=self.suggestion_config['l2_reg'],
+                entropy_weight=self.suggestion_config['entropy_weight'],
+                bl_dec=self.suggestion_config['baseline_decay'],
+                optim_algo=self.suggestion_config['optimizer'],
+                skip_target=self.suggestion_config['skip-target'],
+                skip_weight=self.suggestion_config['skip-weight'],
+                name="Ctrl_"+self.current_study_id)
+
+            self.controller.build_trainer()
+
+        print("Suggestion for StudyJob {} has been initialized.".format(request.study_id))
 
     def GetSuggestions(self, request, context):
+        if request.study_id != self.current_study_id:
```

Comment:
A user may submit two studyjob both with nasrl suggestion, and the two studyjob may call GetSuggestions alternately. It seems NasrlService can only serve one studyjob at same time. right?

Reply:
Could I make the support for multiple studyjobs a separate PR?

Reply:
sure, please create an new issue to trace it